v4/supervisor.go

```diff
@@ -0,0 +1,956 @@
+// participate in the log function propagation and recursive
+// UnstoppedService report.
+//
+// It is legal for GetSupervisor to potentiall return nil, in which case
+// the supervisor-specific behaviors will simply be ignored.
+type HasSupervisor interface {
+	GetSupervisor() *Supervisor
+}
+
+func (s *Supervisor) GetSupervisor() *Supervisor {
+	return s
+}
+
+/*
+Add adds a service to this supervisor.
+
+If the supervisor is currently running, the service will be started
+immediately. If the supervisor has not been started yet, the service
+will be started when the supervisor is. If the supervisor was already stopped,
+this is a no-op returning an empty service-token.
+
+The returned ServiceID may be passed to the Remove method of the Supervisor
+to terminate the service.
+
+As a special behavior, if the service added is itself a supervisor, the
+supervisor being added will copy the Log function from the Supervisor it
+is being added to. This allows factoring out providing a Supervisor
+from its logging. This unconditionally overwrites the child Supervisor's
+logging functions.
+
+*/
+func (s *Supervisor) Add(service Service) ServiceToken {
+	if s == nil {
+		panic("can't add service to nil *suture.Supervisor")
+	}
+
+	if hasSupervisor, isHaveSupervisor := service.(HasSupervisor); isHaveSupervisor {
+		supervisor := hasSupervisor.GetSupervisor()
+		if supervisor != nil {
+			supervisor.spec.LogBadStop = s.spec.LogBadStop
+			supervisor.spec.LogFailure = s.spec.LogFailure
+			supervisor.spec.LogBackoff = s.spec.LogBackoff
+		}
+	}
+
+	s.m.Lock()
+	if s.state == notRunning {
+		id := s.serviceCounter
+		s.serviceCounter++
+
+		s.services[id] = serviceWithName{service, serviceName(service)}
+		s.restartQueue = append(s.restartQueue, id)
+
+		s.m.Unlock()
+		return ServiceToken{uint64(s.id)<<32 | uint64(id)}
+	}
+	s.m.Unlock()
+
+	response := make(chan serviceID)
+	if !s.sendControl(addService{service, serviceName(service), response}) {
+		return ServiceToken{}
+	}
+	return ServiceToken{uint64(s.id)<<32 | uint64(<-response)}
+}
+
+// ServeBackground starts running a supervisor in its own goroutine. When
+// this method returns, the supervisor is guaranteed to be in a running state.
+func (s *Supervisor) ServeBackground(ctx context.Context) {
+	go s.Serve(ctx)
+	s.sync()
+}
+
+/*
+Serve starts the supervisor. You should call this on the top-level supervisor,
+but nothing else.
+*/
+func (s *Supervisor) Serve(ctx context.Context) error {
+	// context documentation suggests that it is legal for functions to
+	// take nil contexts, it's user's responsibility to never pass them in.
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	if s == nil {
+		panic("Can't serve with a nil *suture.Supervisor")
+	}
+	// Take a separate cancellation function so this tree can be
+	// indepedently cancelled.
+	ctx, myCancel := context.WithCancel(ctx)
+	s.ctxMutex.Lock()
+	s.ctx = ctx
+	s.ctxMutex.Unlock()
```

Comment:
I think in all cases, the only use of `s.ctx` is just accessing it's `Done()`, so perhaps you don't need to store the whole context, and could just store the done channel.

Also, select on a nil channel never "fires", so I think it might be safe to not even have a mutex for that?